File: July2021/C++/Find K Closest Elements.cpp
//explanaiton
//https://leetcode.com/explore/challenge/card/july-leetcoding-challenge-2021/608/week-1-july-1st-july-7th/3800/discuss/106426/JavaC++Python-Binary-Search-O(log(N-K)-+-K)

class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int low=0, high=arr.size()-k;
        while(low<high)
        {
            int mid=low+(high-low)/2;
            // cout<<mid<<endl;
            if(x-arr[mid]>arr[mid+k]-x)
                low=mid+1;
            else
                high=mid;
        }
        // cout<<low<<" "<<high<<endl;
        return vector<int>(arr.begin()+low,arr.begin()+low+k);
    }
};

File: July2021/FindKClosestElements.java
class Solution {
    
    // TC : O(n) Linear Search
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> ans = new ArrayList<>();
        int low = 0;
        int high = arr.length-1;
        while(high - low >=k){
            if(Math.abs(arr[low] - x) > Math.abs(arr[high] - x) ){
                low++;
            } else {
                high--;
            }
        }
        
        for(int i=low;i<=high;i++){
            ans.add(arr[i]);   
        }
        return ans;
    }
    
    // Solution 2 : PriorityQueue
     private class Node {
         private int index;
         private int diff;
         Node(int index, int diff){
             this.index= index;
             this.diff = diff;
         }
     }
     public List<Integer> findClosestElements(int[] arr, int k, int x) {
         PriorityQueue<Node> pq = new PriorityQueue<>((a,b) ->((b.diff == a.diff) ? (b.index - a.index) : (b.diff - a.diff)));
        
         for(int i= 0;i<arr.length;i++){
             pq.offer(new Node(i, Math.abs(arr[i] - x)));
             if(pq.size()>k){
                 pq.poll();
             }
         }
        
         List<Integer> ans = new ArrayList<>();
         while(!pq.isEmpty()) {
             ans.add(arr[pq.poll().index]);
         }
        
         Collections.sort(ans);
         return ans;
     }
}